### ISSUE TITLE
Proposal: `generator` keyword as alternative to `function*` syntax

### ISSUE THREAD

**Issue by michalpokusa** [microsoft/TypeScript#55839]
### 🔍 Search Terms

"generator", "keyword", "syntax", "arrow generator", "gen"

### ✅ Viability Checklist

- [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
- [X] This wouldn't change the runtime behavior of existing JavaScript code
- [X] This could be implemented without emitting different JS based on the types of the expressions
- [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
- [X] This feature would agree with the rest of our Design Goals: https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

## TL;DR
Proposed `generator` keyword as alternative to the current `*` symbol. The overall syntax would be more consistent with other keywords like `async`, `static`, `private`, `public`, `readonly` etc. The proposed syntax would also allow arrow functions to be generators. It is backwards compatible and would produce standard JavaScript code for generator while providing better developer experience.

# Introduction

The current TypeScript syntax for generator functions is to use `*` symbol after the `function` keyword. This is the same syntax as in JavaScript.

This proposal is to add a new keyword `generator` that would be used as an alternative to the `*` symbol, but work exactly the same way.

Proposed `generator` keyword in in fact very similar to `async` keyword in the sense that it is a modifier for a function. The `async` keyword is used to mark a function as an async function, and the `generator` keyword would be used to mark a function as a generator function.

It is also worth noting that using both `async` and `generator` keywords on the same function would be possible, and would result in a function that is both async and a generator (examples later).

There is already a (stage 1) proposal for this in ECMAScript:
https://www.proposals.es/proposals/Generator%20arrow%20functions

Additional pros of the proposed syntax:
- It is backwards compatible with existing code, as the `*` symbol would still be supported.
- It would allow arrow functions to be generators (not exactly, more on that later).
- It is more similar to the existing TS syntax.
- It does not change any underlying TS concepts, works like `class` keyword and objects. (Not sure whether this violates "This isn't a runtime feature" in ✅ Viability Checklist, as the transpiled JS code is stil the same)
- It is easier to read and understand, and harder to miss while reading the code.

Potential cons of the proposed syntax:
- `generator` is a longer keyword than `*` symbol (but has the same length as existing `interface` keyword).
- Two ways of doing the same thing (using `*` or `generator` keyword) could be confusing for some developers.

## Possible implementation

Implementing described feature would really come down to adding a new keyword to the TypeScript compiler.

The transpilation would work like this:

- `function* foo() { ... }` would be transpiled to `function* foo() { ... }`
- `generator function foo() { ... }` would be transpiled to `function* foo() { ... }`

in one specific case, the transpiler would have to do a little bit more work:
- `generator () => { ... }` would be transpiled to `function* () { ... }`, as arrow functions cannot be generators in JavaScript, (maybe there could be some other way, that would preserve the `this`)

### 📃 Motivating Example

Introducing the generator keyword as an alternative to the * symbol for defining generator functions in TypeScript offers improved code consistency and readability. This change aligns TypeScript more closely with existing keywords like async, enhancing the language's overall coherence. Additionally, it maintains backward compatibility with existing code, provides the possibility of arrow functions as generator, and simplifies the syntax for developers, making it less error-prone and easier to read.

## Examples

This is the most basic example of a generator function:

```ts
// with "*"
function* foo() {
    ...
}

// with "generator"
generator function foo() {
    ...
}
```
#### Annonymous functions

```ts
// with "*"
const foo = function*() {
    ...
}

// with "generator"
const foo = generator function() {
    ...
}
```

#### Arrow functions

Arrow functions cannot be generators using the current TypeScript and JavaScript syntax. Attempting to add * after an arrow function would not work.

Using the proposed syntax, arrow functions could be generators:

```ts
// with "*"
// (impossible)

// with "generator"
const foo = generator () => {
    ...
}
```

It is important to mention that this **could not be transpiled to standard JavaScript, as arrow functions cannot be generators** in JavaScript. However, it would be possible to transpile it to a standard function, and then add the `*` symbol to it.

```ts
// transpiled to
const foo = function*() {
    ...
}
```

#### Class methods

Proposed keyword would also fit very well with existing class specific keyword like `static`, `private`, `public`, `readonly` etc.

Using the proposed syntax, the class definition would look more unified, as currently the `*` is the only symbol symbol modifier, while all other are keywords.

```ts
class Foo {

    // with "*"
    *bar() {
        ...
    }

    // with "generator"
    generator bar() {
        ...
    }
}
```

It would be important to use keywords in correct order, as **the following would NOT work**:

```ts
class Foo {
    generator static bar() { // <-- WRONG: "generator" before "static"
        ...
    }

    generator async private baz() { // <-- WRONG: "generator" before "async" and "private"
        ...
    }
}
```

Ordering the keywords correctly this way, and only this way, would work. This also preserves the current behavior of TypeScript (`async` cannot be used before `static` or `private` etc.).

```ts
class Foo {
    static generator bar() {
        ...
    }

    private async generator baz() {
        ...
    }
}
```

The same would work very similarly for object methods:

```ts
const foo = {

    // with "*"
    *bar() {
        ...
    }

    // with "generator"
    generator bar() {
        ...
    }
}
```

**Comment by MartinJohns** [comment]
This is explicitly declined in the viability checklist:

> This isn't a runtime feature (e.g. library functionality, **non-ECMAScript syntax with JavaScript output**, new syntax sugar for JS, etc.)

**Comment by michalpokusa** [comment]
> This is explicitly declined in the viability checklist:
> 
> > This isn't a runtime feature (e.g. library functionality, **non-ECMAScript syntax with JavaScript output**, new syntax sugar for JS, etc.)

Thank you for clarifying, I had read them, but as I mentioned later in my post I was not sure whether this fulfilled that point, nevertheless I decided to publish this proposal in case I understood it correctly.
